### PR TITLE
[stark] Round-2: SE(3) local-frame coordinate features for wall-shear

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The Transolver model currently uses raw Cartesian coordinates as input features:
- Surface: `[x, y, z, nx, ny, nz, area]` (7D)
- Volume:  `[x, y, z, sdf]` (4D)

These are absolute coordinates in a single global frame. A rigid-body shift or rotation of the car geometry should not change the flow predictions — yet the model must learn this invariance from scratch across only 400 training cases. This is an unnecessary learning burden.

**The fix:** augment the existing features with SE(3)-equivariant relative coordinates derived from the local surface normal frame at each point. We use the **known surface normal vector** (already present in the input as `[nx, ny, nz]`) as the primary frame axis to build an orthonormal local frame. For each surface point, we express its global position in this local frame as 3 additional features. This:

1. Makes the representation partially rotation-equivariant without any architecture change
2. Directly encodes how each point relates to its own surface geometry
3. Should especially help `tau_x`, `tau_y`, `tau_z` — wall shear components transform as pseudo-vectors under frame changes, and encoding them in a local frame reduces the regression burden

**Evidence:** SE(2)-equivariant encoding on TandemFoil (noam branch, PR #2377 MERGED) produced robust improvement across split types. This is the 3D extension of that proven win — feature preprocessing only, zero architecture changes, very low risk.

## Instructions

In `target/train.py`, add `--use-local-frame-features` (default off) that augments input features with local-frame coordinates.

**1. Add CLI argument:**
```python
parser.add_argument("--use-local-frame-features", action="store_true", default=False,
    help="Augment surface/volume features with SE(3) local-frame coordinates (+3 dims each)")
parser.add_argument("--no-use-local-frame-features", dest="use_local_frame_features", action="store_false")
```

**2. Add the feature augmentation function (module-level, before training loop):**
```python
def add_local_frame_features(surface_x, volume_x):
    """
    Augment surface/volume features with SE(3)-equivariant local-frame coords.
    Surface: [B, N_s, 7] -> [B, N_s, 10]  (xyz in local normal frame appended)
    Volume:  [B, N_v, 4] -> [B, N_v, 7]   (xyz relative to surface centroid in mean frame)
    """
    import torch.nn.functional as F

    # --- Surface local frame from normal ---
    xyz_s = surface_x[..., :3]   # [B, N_s, 3]
    nml   = surface_x[..., 3:6]  # [B, N_s, 3] surface normals (already unit)

    # Build orthonormal frame {n, t1, t2}. Use y-axis as reference.
    ref = torch.zeros_like(nml); ref[..., 1] = 1.0
    parallel = (nml[..., 1].abs() > 0.9)  # handle n nearly parallel to y
    ref[parallel, 1] = 0.0; ref[parallel, 2] = 1.0
    t1 = F.normalize(torch.cross(nml, ref, dim=-1), dim=-1)
    t2 = torch.cross(nml, t1, dim=-1)

    local_n  = (xyz_s * nml).sum(-1, keepdim=True)
    local_t1 = (xyz_s * t1).sum(-1, keepdim=True)
    local_t2 = (xyz_s * t2).sum(-1, keepdim=True)
    surface_x_aug = torch.cat([surface_x, local_n, local_t1, local_t2], dim=-1)  # [B,N_s,10]

    # --- Volume: relative to surface centroid in mean-normal frame ---
    mean_xyz = xyz_s.mean(dim=1, keepdim=True)
    mean_nml = F.normalize(nml.mean(dim=1, keepdim=True), dim=-1)
    ref_v = torch.zeros_like(mean_nml); ref_v[..., 1] = 1.0
    par_v = (mean_nml[..., 1].abs() > 0.9)
    ref_v[par_v, 0, 1] = 0.0; ref_v[par_v, 0, 2] = 1.0
    mt1 = F.normalize(torch.cross(mean_nml, ref_v, dim=-1), dim=-1)
    mt2 = torch.cross(mean_nml, mt1, dim=-1)

    rel_v = volume_x[..., :3] - mean_xyz
    vol_n  = (rel_v * mean_nml).sum(-1, keepdim=True)
    vol_t1 = (rel_v * mt1).sum(-1, keepdim=True)
    vol_t2 = (rel_v * mt2).sum(-1, keepdim=True)
    volume_x_aug = torch.cat([volume_x, vol_n, vol_t1, vol_t2], dim=-1)  # [B,N_v,7]

    return surface_x_aug, volume_x_aug
```

**3. Apply the augmentation just before the forward pass:**
```python
if config.use_local_frame_features:
    surface_x_batch, volume_x_batch = add_local_frame_features(surface_x_batch, volume_x_batch)
```

**4. Model input dimension adjustment:**
The model's surface input dim changes from 7 → 10 and volume from 4 → 7. Find where the model is initialized in `train.py` and update accordingly:
```python
surface_in_dim = 10 if config.use_local_frame_features else 7
volume_in_dim  = 7  if config.use_local_frame_features else 4
```
Pass these to the Transolver constructor. If the constructor argument is named something like `in_channels`, `input_dim`, `surf_dim`, or is set via config fields, update appropriately. If the dim is hardcoded internally, the simplest fix is a learned linear layer that projects 10D→7D before the model input while keeping the architecture unchanged.

**5. Run command:**
```bash
cd target/
python train.py \
  --volume-loss-weight 2.0 \
  --batch-size 8 \
  --validation-every 1 \
  --lr 2e-4 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --ema-decay 0.9995 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms \
  --use-local-frame-features \
  --wandb-group round2-se3-local-frame
```

**Diagnostic:** Focus on `test_primary/wall_shear_*_rel_l2_pct` — the hypothesis predicts the largest gains on the directional wall-shear axes. If all three axes improve, the equivariant encoding is working.

## Baseline (current yi best — PR #9, run y2gigs61)

| Metric | Current best | AB-UPT target |
|---|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | **17.3933** | — |
| `surface_pressure_rel_l2_pct` | **11.0733** | 3.82 |
| `wall_shear_rel_l2_pct` | **18.3180** | 7.29 |
| `volume_pressure_rel_l2_pct` | **15.2059** | 6.08 |
| `wall_shear_x_rel_l2_pct` | **15.6465** | 5.35 |
| `wall_shear_y_rel_l2_pct` | **21.8605** | 3.65 |
| `wall_shear_z_rel_l2_pct` | **23.1803** | 3.63 |

Beat `abupt_axis_mean_rel_l2_pct < 17.39` to set a new baseline. Note: FiLM PR #8 (frieren) is pending merge with 16.53 — so the effective bar will be **16.53** once merged.

## Reproduce baseline
```bash
cd target/
python train.py \
  --volume-loss-weight 2.0 \
  --batch-size 8 \
  --validation-every 1 \
  --lr 2e-4 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --ema-decay 0.9995 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #25 because ADVISOR: Closing this PR — `stark` is not a member of this research cohort. The 16 students on this track are: alphonse, askeladd, chihiro, edward, emma, fern, frieren, gilbert, haku, kohaku, nezuko, norman, senku, tanjiro, thorfinn, violet.

The SE(3) local-frame coordinate-features hypothesis itself is sound — it is one of the top-priority Round-2 candidates (A02 in `research/RESEARCH_IDEAS_2026-04-28_ROUND2_ARCHITECTURES.md`). It will be reassigned to an actual idle student in a future round.

Apologies for the confusion — this assignment was created in error.

— ADVISOR.
